src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -443,178 +286,157 @@ static smart_amp_proc get_smart_amp_process(struct comp_dev *dev,
 	}
 }
 
-static int smart_amp_copy(struct comp_dev *dev)
+/* Only reset the src_input and fb_input if the num_input_buffers
+ * (get from module adapter)has changed.
+ */
+static void smart_amp_input_setup(struct processing_module *mod,
+				  struct input_stream_buffer *inputs,
+				  int32_t num_input_buffers)
+{
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	int i;
+
+	if (sad->input_stream_num != num_input_buffers) {
+		if (num_input_buffers == 1) {
+			sad->src_input = &inputs[0];
+			sad->fb_input = NULL;
+		} else {
+			for (i = 0; i < num_input_buffers; i++) {
+				if (IPC4_SINK_QUEUE_ID(inputs[i].comp_buf_id) ==
+				   SOF_SMART_AMP_FEEDBACK_QUEUE_ID)
+					sad->fb_input = &inputs[i];
+				else
+					sad->src_input = &inputs[i];
+			}
+		}
+
+		sad->input_stream_num = num_input_buffers;
+	}
+}
+
+static int smart_amp_process(struct processing_module *mod,
+			     struct input_stream_buffer *input_buffers, int num_input_buffers,
+			     struct output_stream_buffer *output_buffers, int num_output_buffers)
 {
-	struct smart_amp_data *sad = comp_get_drvdata(dev);
-	struct comp_buffer __sparse_cache *source_buf = buffer_acquire(sad->source_buf);
-	struct comp_buffer __sparse_cache *sink_buf = buffer_acquire(sad->sink_buf);
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer __sparse_cache *fb_buf;
+	struct comp_buffer __sparse_cache *src_buf;
 	uint32_t avail_passthrough_frames;
-	uint32_t avail_feedback_frames;
 	uint32_t avail_frames = 0;
-	uint32_t source_bytes;
 	uint32_t sink_bytes;
-	uint32_t feedback_bytes;
-
-	comp_dbg(dev, "smart_amp_copy()");
-
-	/* available bytes and samples calculation */
-	avail_passthrough_frames =
-		audio_stream_avail_frames(&source_buf->stream,
-					  &sink_buf->stream);
+	uint32_t fb_bytes;
+	uint32_t source_bytes;
 
-	k_mutex_lock(&sad->lock, K_FOREVER);
```

Comment:
why do you think you don't need locking any more? The module adapter API doesn't lock internally

Reply:
Rander told me the sad->lock is just for bind/unbind, and suggest me remove it .
https://github.com/thesofproject/sof/pull/7424#discussion_r1161513266.

Reply:
yes, this is lock is used for bind/unbind thread and process thread. Now bind/unbind are removed, all lock reference is used in the same thread so we don't need it now.

Reply:
it's also used in `.prepare()`, isn't it called from the IPC context?

Reply:
@RanderWang ok, so we're relying on IPC serialisation. But then that also holds for `.bind()` and `.unbind()`, including for IPC3, doesn't it? So we can remove it from the original smart-amp test, before this PR? Can we make that a separate commit in this PR at least? That seems important enough to me to be just done as a side effect in one of commits.

Reply:
@lyakh  No, .bind() and .unbind() is not dependent on IPC serialisation. Since .bind & .unbind are triggered by another stream. No such issue on IPC3 since it requires that feedback stream should be started before smart amp stream please check smart_amp_prepare